fix(ivy): ngOnChanges only runs for binding updates

benlesh opened the pull request:
Updates ivy to trigger onChanges only when bindings update on not when properties are set.

- Currently there are still test failures in a few places, I'm not sure they are related as `core` is passing, if they are not, I'll mark them `fixMeIvy` and move on, because this PR is already pretty large
- SimpleChanges is now stored alongside the directive instance in LView. May changes had to be made to accommodate looking up directive and component instances from LView because of this.
- refactors hooks.ts slightly
- Adds handling for `onChanges` events to `hooks.ts` in `callHooks`.
- Fixes a bug where ngBaseDef creation was not honoring pre-existing ngComponentDef and ngDirectiveDef items
- Updates the Inherit Definition Feature to support inheriting onChanges
- Removes NgOnChangesFeature

### Addendum

- Also solves an issue where `ts_web_test_suite` rule was sometimes adding `init_browser_spec.js` out of order, causing Karma tests to fail when examining uninitialized `isNode` variable in global scope.
- Had to increase the payload limit of hello-world to `187437` from `185238`

Commit:
docs: fix typo in TOH http section (#26127)

PR Close #26127

alan-agius4 committed:
docs: add more info on what is being cached (#27343)

PR Close #27343

RGelman committed:
docs: add missing comma to environment config in build guide (#27476)

PR Close #27476

AndrewKushnir committed:
fix(ivy): implement 'TestBed.overrideTemplateUsingTestingModule' function (#27717)

Adding 'TestBed.overrideTemplateUsingTestingModule' function that overrides Component template and uses TestingModule as a scope.

PR Close #27717

AndrewKushnir committed:
fix(ivy): resolve `forwardRefs` correctly in TestBed (JIT) (#27737)

Forward refs in some places (like imports/export/providers/viewProviders/queries) were not resolved before passing to compilation phase. Now we resolve missing refs before invoking compile function.

PR Close #27737

gkalpak committed:
fix(ivy): ensure unique template and listener function identifiers (#27766)

Previously, there could be identical template/listener function names
for a component's template, if it had multiple similarly structured
nested sub-templates or listeners.
This resulted in build errors:
`Identifier '<SOME_IDENTIFIER>' has already been declared`

This commit fixes this by ensuring that the template index is included
in the `contextName` passed to the `TemplateDefinitionBuilder`
responsible for processing nested sub-templates.
Similarly, the template or element index is included in the listener
names.

PR Close #27766

kara committed:
fix(ivy): ng-content tags in re-inserted templates should walk declaration tree (#27783)

This PR assures that content projection works if an <ng-content> tag is
placed inside an <ng-template> in one component and that <ng-template>
is inserted into a different component. It fixes a bug where the
projection instruction code would walk up the insertion tree to find
selector data instead of the declaration tree.

PR Close #27783

benlesh committed:
fix(ivy): ngOnChanges only runs for binding updates

benlesh committed:
fix(ivy): ngOnChanges only runs for binding updates

benlesh commented on packages/common/src/directives/ng_template_outlet.ts:
```diff
@@ -51,6 +48,7 @@ export class NgTemplateOutlet implements OnChanges {
     if (recreateView) {
       if (this._viewRef) {
         this._viewContainerRef.remove(this._viewContainerRef.indexOf(this._viewRef));
+        // this._viewRef = null;
```
I'll revert this.

benlesh commented on packages/core/src/di/injector.ts:
```diff
@@ -352,6 +352,7 @@ function computeDeps(provider: StaticProvider): DependencyRecord[] {
   let deps: DependencyRecord[] = EMPTY;
   const providerDeps: any[] =
       (provider as ExistingProvider & StaticClassProvider & ConstructorProvider).deps;
+  debugger;
```
I'll revert this.

benlesh commented on packages/core/src/render3/definition.ts:
```diff
@@ -249,6 +250,7 @@ export function defineComponent<T>(componentDefinition: {
   const type = componentDefinition.type;
   const typePrototype = type.prototype;
   const declaredInputs: {[key: string]: string} = {} as any;
+
```
Will revert.

mary-poppins commented:
You can preview fad83a8 at https://pr27787-fad83a8.ngbuilds.io/.

benlesh commented on packages/core/src/render3/hooks.ts:
```diff
@@ -9,26 +9,36 @@
-
+import {FLAGS, HookData, LView, LViewFlags, TView } from './interfaces/view';
 
+const ON_CHANGES = -2;
```
Will add doc comment. 

NOTE: I chose -2 because the array this goes in stored numbers that were all indices, and -1 generally signals "not found in array". This is being used as a marker in the array, not an index. I could use something else, just figured this was lightweight.

benlesh commented on packages/core/src/render3/hooks.ts:
```diff
@@ -9,26 +9,36 @@
  */
-export function queueInitHooks(
-    index: number, onInit: (() => void) | null, doCheck: (() => void) | null, tView: TView): void {
+export function setupHooksDirectiveStart(
```
Renamed purely for clarity around what this was doing. `queueInitHooks` was confusing to me for a while when I started working on this.

benlesh commented on packages/core/src/render3/hooks.ts:
```diff
@@ -9,26 +9,36 @@
+
+  if (onChanges) {
+    (tView.initHooks || (tView.initHooks = [])).push(index, ON_CHANGES, onChanges);
+    (tView.checkHooks || (tView.checkHooks = [])).push(index, ON_CHANGES, onChanges);
```
Note the ON_CHANGES marker added to the array.

benlesh commented on packages/core/src/render3/hooks.ts:
```diff
@@ -119,13 +140,32 @@ export function executeHooks(
+    const record = currentView[arr[i] as number];
+    const hasOnChanges = Array.isArray(record);
+    let inst = hasOnChanges ? record[0] : record;
+    const isOnChanges = arr[i + 1] === ON_CHANGES;
```
Down here is where we're looking for the ON_CHANGES marker as it will cause us to pull out the SimpleChangse ref and adjust our "step" in the loop.

benlesh commented on packages/core/src/render3/instructions.ts:
```diff
@@ -20,7 +20,7 @@ import {bindingUpdated, bindingUpdated2, bindingUpdated3, bindingUpdated4} from
 import {diPublicInInjector, getNodeInjectable, getOrCreateInjectable, getOrCreateNodeInjectorForNode, injectAttributeImpl} from './di';
 import {throwMultipleComponentError} from './errors';
-import {executeHooks, executeInitHooks, queueInitHooks, queueLifecycleHooks} from './hooks';
+import {executeHooks as executeHooks, setupHooksDirectiveStart, setupHooksDirectiveEnd, executeInitHooks} from './hooks';
```
Will revert `executeHooks as executeHooks`, I think VS Code did that.

benlesh commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1032,10 +1047,30 @@ export function createTNode(
+      directiveInstance = directiveInstance[0];
+      const directiveDef = lView[TVIEW].data[directiveIndex] as DirectiveDef<any>;
+      const declaredName = directiveDef.declaredInputs[propName];
+      const firstChange = !hasChanged.has(privateName);
```
The use of a `Set` here is debatable. Happy to use something else. But we can probably refine that in another PR.

benlesh commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1514,11 +1549,21 @@ function instantiateAllDirectives(tView: TView, lView: LView, tNode: TNode) {
+      const inst = lView[i];
+      // this is [instance, SimpleChanges|null, hasChangedLookup]
+      lView[i] = [inst, null, new Set<string>()];
+    }
```
This is where we're storing the instance, the `SimpleChanges` (or null), and the `Set` of what we've already seen in the `LView`.

benlesh commented on packages/core/test/linker/change_detection_integration_spec.ts:
```diff
@@ -1707,6 +1708,7 @@ function patchLoggingRenderer2(rendererFactory: RendererFactory2, log: RenderLog
     const origSetProperty = renderer.setProperty;
     const origSetValue = renderer.setValue;
     renderer.setProperty = function(el: any, name: string, value: any): void {
+      debugger;
```
will revert

benlesh commented on packages/core/test/linker/change_detection_integration_spec.ts:
```diff
@@ -1716,6 +1718,7 @@ function patchLoggingRenderer2(rendererFactory: RendererFactory2, log: RenderLog
       }
       origSetValue.call(renderer, node, value);
     };
+    debugger;
```
will revert

benlesh commented on packages/core/test/render3/inherit_definition_feature_spec.ts:
```diff
@@ -493,6 +493,7 @@ describe('InheritDefinitionFeature', () => {
     let subDir !: SubDirective;
 
     class SuperDirective {
+      @Input()
```
Will revert.

benlesh commented on packages/core/test/render3/inherit_definition_feature_spec.ts:
```diff
@@ -6,8 +6,8 @@
-import {Inject, InjectionToken} from '../../src/core';
-import {ComponentDef, DirectiveDef, InheritDefinitionFeature, NgOnChangesFeature, ProvidersFeature, RenderFlags, allocHostVars, bind, defineBase, defineComponent, defineDirective, directiveInject, element, elementProperty, load} from '../../src/render3/index';
+import {Inject, InjectionToken, Input} from '../../src/core';
+import {ComponentDef, DirectiveDef, InheritDefinitionFeature, ProvidersFeature, RenderFlags, allocHostVars, bind, defineBase, defineComponent, defineDirective, directiveInject, element, elementProperty, load, elementEnd} from '../../src/render3/index';
```
Will revert.

benlesh commented on packages/core/test/render3/inherit_definition_feature_spec.ts:
```diff
@@ -6,8 +6,8 @@
 
-import {Inject, InjectionToken} from '../../src/core';
-import {ComponentDef, DirectiveDef, InheritDefinitionFeature, NgOnChangesFeature, ProvidersFeature, RenderFlags, allocHostVars, bind, defineBase, defineComponent, defineDirective, directiveInject, element, elementProperty, load} from '../../src/render3/index';
+import {Inject, InjectionToken, Input} from '../../src/core';
```
will revert

benlesh commented on packages/core/test/render3/lifecycle_spec.ts:
```diff
@@ -2091,6 +2089,7 @@ describe('lifecycles', () => {
       }, 2, 4, defs);
 
       const fixture = new ComponentFixture(App);
+
```
will revert.

benlesh commented on packages/core/test/render3/lifecycle_spec.ts:
```diff
@@ -2419,6 +2417,7 @@ describe('lifecycles', () => {
       }, 2, 0, [Comp]);
 
       const fixture = new ComponentFixture(App);
+
```
will revert.

benlesh commented on packages/core/test/render3/query_spec.ts:
```diff
@@ -1698,7 +1698,11 @@ describe('query', () => {
               }
               if (rf & RenderFlags.Update) {
-                queryRefresh(tmp = load<QueryList<any>>(0)) &&
+                tmp = load<QueryList<any>>(0);
```
will revert.

benlesh commented on packages/core/test/render3/view_container_ref_spec.ts:
```diff
@@ -1631,7 +1631,7 @@ describe('ViewContainerRef', () => {
           }
         },
-        features: [NgOnChangesFeature],
+
```
will revert.

JoostK commented on packages/core/src/render3/hooks.ts:
```diff
@@ -101,10 +115,17 @@ export function executeInitHooks(
 }
 
+
+// TODO(benlesh): It seems like we could just replace this with some ternaries and
```
Yes, we can: #25295. That PR got lost and is quite outdated now, I can update it after this one lands if that's desirable.

benlesh replied:
Now that this method is actually checking lView flags, I don't think we can. I'll remove the comment.

gkalpak committed:
refactor(ivy): minor refactoring / code clean-up (#27766)

PR Close #27766

marclaval committed:
test(ivy): add root cause analysis for failing core tests (view injector integration) (#27912)

PR Close #27912

benlesh commented on packages/core/src/render3/hooks.ts:
```diff
@@ -101,10 +115,17 @@ export function executeInitHooks(
+ * @param creationMode Whether or not we're running in creationMode.
  */
 export function executeHooks(
     currentView: LView, allHooks: HookData | null, checkHooks: HookData | null,
```
Oh... good catch, I think that got trampled when I rebased.

mhevery committed:
fix(ivy): don’t publish animation bindings as attributes (#27805)

Some of the animation tests have been failing because animation gets
triggered multiple times. The reason for this is that the compiler was
generating static attribute bindings in addition to dynamic bindings.
This created multiple writes to the animation render which failed the
tests.

PR Close #27805

Commit:
fix(forms): match getError and hasError to get method signature (#20211)

Internally getError and hasError call the AbstractControl#get method which takes  `path: Array<string | number> | string` as input, since there are different ways to traverse the AbstractControl tree.
This change matches the method signitures of all methods that use this.

PR Close #20211

msal4 committed:
docs: fix npm description in architecture modules guide (#27818)

PR Close #27818

IgorMinar committed:
test: remove bower and polymer benchmarks (#27931)

the polymer benchmarks are super old and not relevant any more

and these benchmarks were the only reason why we needed bower at all

so long, bower. thanks for all the fish.

PR Close #27931

alan-agius4 committed:
docs: improve description of bundle budgets (#27833)

PR Close #27833

benlesh committed:
feat(ivy): properly handle inheriting lifecycle hooks from plain classes

benlesh committed:
feat(ivy): properly handle inheriting lifecycle hooks from plain classes

devversion committed:
fix(bazel): unable to launch protractor test on windows (#27850)

Due to an incorrect environment variable name, it's currently not possible to launch Protractor on Windows using the Bazel protractor rule.

PR Close #27850

crisbeto committed:
test(ivy): update test failure root cause (#27855)

Updates the root cause for a couple of the test failures.

PR Close #27855

kara committed:
docs(forms): update desc for hasError and getError (#27861)

This commit adds docs for the changes made in #20211.

Closes #19734.

PR Close #27861

petebacondarwin committed:
fix(docs-infra): upgrade to latest dgeni-packages to fix linking problem (#27864)

This new version of dgeni-packages gives the main (implemented)
overload of a method the correct id and aliases, which allow it to be
automatically linked.

See https://github.com/angular/dgeni-packages/commit/398f35da303a839525cdb5aa690de394e905b1d5

Fixes #27820
Closes #27821

PR Close #27864

cyrilletuzi committed:
docs(service-worker): update service worker configuration doc (#27868)

PR Close #27868

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview 3bc95f8 at https://pr27787-3bc95f8.ngbuilds.io/.

crisbeto committed:
fix(ivy): ngtsc program emit ignoring custom transformers (#27837)

Fixes the `customTransformers` that are passed to the `NgtscProgram.emit` not being passed along.

PR Close #27837

gkalpak committed:
test(docs-infra): increase timeout for redirection to external URL (#27903)

Occasionally, external URLs take longer to load, which causes CI flakes.

PR Close #27903

marclaval committed:
fix(ivy): pipes should overwrite pipes with later entry in the pipes array (#27910)

PR Close #27910

marclaval committed:
test(ivy): enable more @angular/core tests (#27912)

PR Close #27912

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview 27bd171 at https://pr27787-27bd171.ngbuilds.io/.

mary-poppins commented:
You can preview feb69e4 at https://pr27787-feb69e4.ngbuilds.io/.

AndrewKushnir committed:
fix(ivy): support bindings in ICUs in nested contexts (#27914)

Prior to this change, ICU extraction logic was not taking into account nested bindings (that look like this: �0:1�) and only accounted for top level bindings (like this �0�). As a result, ICUs were not parsed and remained as text in the output. Now the extraction logic (regular expressions) take into account the nested bindings format as well.

PR Close #27914

mhevery commented on packages/core/test/linker/change_detection_integration_spec.ts:
```diff
@@ -775,7 +775,7 @@ const TEST_COMPILER_PROVIDERS: Provider[] = [
       describe('ngOnInit', () => {
         it('should be called after ngOnChanges', fakeAsync(() => {
-             const ctx = createCompFixture('<div testDirective="dir"></div>');
+             const ctx = createCompFixture('<div testDirective="dir" [a]="\'test\'" [b]="\'test\'"></div>');
```
Why did you have to change this test?

benlesh replied:
Ah.. this was changed before I realized I needed up update SimpleChanges in setInputsFromAttrs

Reverted.

mhevery commented on packages/core/test/render3/inherit_definition_feature_spec.ts:
```diff
@@ -519,6 +518,9 @@ describe('InheritDefinitionFeature', () => {
         element(0, 'div', ['subDir', '']);
       }
+      if (rf & RenderFlags.Update) {
+        elementProperty(0, 'someInput', bind('Weee'));
```
You added conditions to test, but you did not add any assertions. Either revert or add assertions.

benlesh replied:
The assertion is below. This test used to rely on the fact we were patching the setters, but now binding is required to trigger onChanges.

mhevery commented on packages/core/test/view/provider_spec.ts:
```diff
@@ -137,34 +138,35 @@ import {ARG_TYPE_VALUES, checkNodeInlineOrDynamic, createRootView, createAndGetR
-                  '  StaticInjectorError(Platform: core)[SomeService -> Dep]: \n' +
-                  '    NullInjectorError: No provider for Dep!');
-        });
+        fixmeIvy('FW-875: The source of the error is missing in the `StaticInjectorError` message')
```
Why is code which used to pass now failing?

benlesh replied:
Possibly some poorly resolved merge conflict. It doesn't seem to be related to any code touched for onChanges. I think @ocombe might be able to provide insight here, as I'm pretty sure he's the one who I saw fix this issue.

ocombe replied:
I think it's @pkozlowski-opensource that fixed this one

mhevery committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mhevery reviewed (changes requested):
I have created a fixup commit 2942cb3a69. Can we get this PR green, so that I can take another look?

mary-poppins commented:
You can preview 2942cb3 at https://pr27787-2942cb3.ngbuilds.io/.

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview b69a64a at https://pr27787-b69a64a.ngbuilds.io/.

benlesh committed:
test(ivy): correct attribute marker in broken test (#27919)

Test incorrectly marks the attribute as a 'Class', when it should be 'SelectOnly'

PR Close #27919

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview 41cfe45 at https://pr27787-41cfe45.ngbuilds.io/.

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview a514402 at https://pr27787-a514402.ngbuilds.io/.

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview 774da4c at https://pr27787-774da4c.ngbuilds.io/.

mary-poppins commented:
You can preview 2d2fdd2 at https://pr27787-2d2fdd2.ngbuilds.io/.

benlesh committed:
fix(bazel): ts_web_test_suite now properly includes init_browser_spec.js

There was an issue where init_browser_spec.js was being run out of order, this change adds it as a runtime dependency so it is executed when it needs to be preventing tests from bombing when they try to read from isNode from global scope before it is defined

benlesh committed:
fix(bazel): ts_web_test_suite now properly includes init_browser_spec.js

There was an issue where init_browser_spec.js was being run out of order, this change adds it as a runtime dependency so it is executed when it needs to be preventing tests from bombing when they try to read from isNode from global scope before it is defined

mary-poppins commented:
You can preview 1efffdd at https://pr27787-1efffdd.ngbuilds.io/.

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview cf67e48 at https://pr27787-cf67e48.ngbuilds.io/.

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

benlesh commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1736,16 +1773,28 @@ function addComponentLogic<T>(
+    const hasOnChanges = isOnChangesDirectiveWrapper(record);
+    const instance = hasOnChanges ? record.instance : record;
+
+    for (let i = 0; i < initialInputs.length; i += 3) {
```
You've changed this in your commit.

matsko committed:
fix(ivy): ensure @animation host bindings/listeners work properly (#27896)

PR Close #27896

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview dcb214c at https://pr27787-dcb214c.ngbuilds.io/.

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview d312aa3 at https://pr27787-d312aa3.ngbuilds.io/.

marclaval committed:
fix(ivy): element injector on the host of a component should see viewProviders of that component (#27928)

PR Close #27928

gkalpak committed:
ci: make `integration_test` job logs less verbose (#27934)

The build and test progress logs make the CI log output so long that it
can't be displayed in the UI and one has to download and view the file
locally instead. This makes it harder to get to the interesting lines,
such as error messages.

Similar to #26869, but for the `bazel-schematics` integration project.

PR Close #27934

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview 483b598 at https://pr27787-483b598.ngbuilds.io/.

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview 53c312f at https://pr27787-53c312f.ngbuilds.io/.

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview 511ece6 at https://pr27787-511ece6.ngbuilds.io/.

benlesh committed:
fixup! fix(ivy): ngOnChanges only runs for binding updates

mary-poppins commented:
You can preview 32bbadc at https://pr27787-32bbadc.ngbuilds.io/.

kara commented:
@benlesh This PR seems to have pulled in a number of unrelated commits from master (like forms changes and animations changes, etc). It's not really reviewable in this state. Can you fix the Github history so it only contains your own commits?

benlesh commented:
@kara looks like a bad rebase. Not sure how that happened, I might just have to start a new branch and cherry-pick my changes over, which would mean a different PR. I'll see what I can do.
